sinon/lib/stub.py

```diff
@@ -17,118 +17,178 @@ class SinonStub(SinonSpy):
     (4) a module
 
     All function/module will be replaced into customized or empty function/class
-    Able to be assigned special condition (on which call, with what args...etc)
+    Able to be assigned special conditions (on which call, with what args, etc.)
     """
 
     def __init__(self, obj=None, prop=None, func=None):
         super(SinonStub, self).__init__(obj, prop)
-        self.stubfunc = func if func else Wrapper.empty_function
-        super(SinonStub, self).wrap2stub(self.stubfunc)
-        # Todo: target is a dirty hack
-        self.condition = {"args":[], "kwargs":[], "action": [], "oncall":[], "target":self.obj}
-        self.cond_args = self.cond_kwargs = self.oncall = None
-
-    def __append_condition(self, func):
-        self.condition["args"].append(self.cond_args)
-        self.condition["kwargs"].append(self.cond_kwargs)
-        self.condition["oncall"].append(self.oncall)
-        self.condition["action"].append(func)
-        self.cond_args = self.cond_kwargs = self.oncall = None
+        self._stubfunc = func if func else Wrapper.empty_function
+        super(SinonStub, self).wrap2stub(self._stubfunc)
+        self._cond_args = self._cond_kwargs = self._oncall = None
+        self._conditions = {"args":[], "kwargs":[], "action": [], "oncall": [], "target": self.obj}            
+
+    def _append_condition(self, sinon_stub_condition, func):
+        '''
+        Permanently saves the current (volatile) conditions, which would be otherwise lost
+
+        Args:
+            sinon_stub_condition: the SinonStubCondition object that holds the current conditions
+            func: returns a value or raises an exception, as specified by the user
+        Returns: the SinonStub._conditions dictionary (for convenience)
+        '''
+        self._conditions["args"].append(sinon_stub_condition._cond_args)
+        self._conditions["kwargs"].append(sinon_stub_condition._cond_kwargs)
+        self._conditions["oncall"].append(sinon_stub_condition._oncall)
+        self._conditions["action"].append(func)
+        return self._conditions
+
+    def _get_original(self):
```

Comment:
I believe this function is unnecessary, you have pass self as a parameter `copy_of=self` while you construct object `SinonStubCondition`.

Reply:
`SinonStubCondition._get_original` is a recursive function, and the version in `SinonStub` is the termination clause. It's more of a style thing, but it seems cleaner to me to do it this way. Otherwise, you are right that I could put all of the logic inside of `SinonStubCondition`.

Now that I've explained my logic, which solution do you prefer?

Reply:
I found a way to remove `_get_original` and the recursion. Please see the new [Pull Request](https://github.com/note35/sinon/pull/34)